openedx/features/enterprise_support/enrollments/utils.py

```diff
@@ -81,6 +67,127 @@ def lms_enroll_user_in_course(
+        try:
+            if current_enrollment:
+                if (
+                    current_enrollment['mode'] == desired_mode
+                    and current_enrollment['is_active'] == is_active
+                ):
+                    return None
+                response = enrollment_api.update_enrollment(
+                    username,
+                    str(course_id),
+                    mode=desired_mode,
+                    is_active=is_active,
+                    enrollment_attributes=None,
+                )
+                if not response or (
+                    response['mode'] != desired_mode or
+                    response['is_active'] != is_active
+                ):
+                    log.exception(
+                        "An error occurred while updating the course enrollment for user "
+                        "[%s]: course run = [%s], enterprise_uuid = [%s], is_active = [%s], ",
+                        username,
+                        course_id,
+                        str(enterprise_uuid),
+                        is_active,
+                    )
+                    raise CourseEnrollmentNotUpdatableError(
+                        f"Unable to upgrade enrollment for user {username} "
+                        "in course {course_id} to {desired_mode} mode."
+                    )
+            else:
+                response = enrollment_api.add_enrollment(
+                    username,
+                    str(course_id),
+                    mode=desired_mode,
+                    is_active=is_active,
+                    enrollment_attributes=None,
+                    enterprise_uuid=enterprise_uuid,
+                )
+                if not response:
+                    log.exception(
+                        "An error occurred while creating the new course enrollment for user "
+                        "[%s] in course run [%s]",
+                        username,
+                        course_id,
+                    )
+                    raise CourseEnrollmentError(
+                        f"Unable to create enrollment for user {username} in course {course_id}."
+                    )
+                return response
+        except CourseEnrollmentExistsError:
+            log.warning(
+                "An enrollment already exists for user [%s] in course run [%s].",
+                username,
+                course_id,
+            )
+            return None
+        except (CourseEnrollmentError, CourseEnrollmentNotUpdatableError) as error:
+            log.exception(
+                "Raising error [%s] for user "
+                "[%s]: course run = [%s], enterprise_uuid = [%s], is_active = [%s], ",
+                error,
+                username,
+                course_id,
+                str(enterprise_uuid),
+                is_active,
+            )
+            raise error
+        finally:
+            # Assumes that the ecommerce service uses an API key to authenticate.
```

Comment:
Nit: maybe make this comment less specific to ecommerce?  Or just kill the comment?